README.md
--- 

# motifmatchr

[![Build Status](https://travis-ci.org/GreenleafLab/motifmatchr.svg?branch=master)](https://travis-ci.org/GreenleafLab/motifmatchr)

## Introduction

motifmatchr is an R package for fast motif matching, using C++ code from the MOODS library. The MOODS library was developed by Pasi Rastas, Janne Korhonen, and Petri Martinmäki. The core C++ library from MOODs version MOODS 1.9.3 code has been included in this repository. 

## Note on recent function name changes

The motifmatchr package recently changed to switch over to camelCase from snake_case.  All exported functions now use camelCase, e.g. `match_motifs` is now `matchMotifs`. If following the current documentation but using an earlier version of the package, either update the package or be aware of the discrepancy. This change was made to comply with Bioconductor naming preferences.    

## Installation

Installation is easiest using the devtools package. The function `install_github` will install the package.

``` r
devtools::install_github("GreenleafLab/motifmatchr")
```

A number of needed packages are installed in this process. One of the dependencies has a system requirement for the gsl library, so if this is not installed already it may need to be installed separately.  

## matchMotifs

The primary method of motifmatchr is `matchMotifs`.  This method has two mandatory arguments:

1) Position weight matrices or position frequency matrices, stored in the PWMatrix, PFMatrix, PWMatrixList, or PFMatrixList objects from the TFBSTools package

2) Either a set of genomic ranges (GenomicRanges or RangedSummarizedExperiment object) or a set of sequences (either DNAStringSet, DNAString, or simple character vector)

If the second argument is a set of genomic ranges, a genome sequence is also required. If the genomic ranges include seqinfo, by default the genome specified in the seqinfo will be used (if the relevant BSgenome package is installed). Otherwise you can supply either a short string specifying the genome build if the corresponding BSgenome object is installed, a BSgenone object, a DNAStringSet object, or a FaFile object pointint to a fasta file.  

The method can return three possible outputs, depending on the `out` argument:

1) (Default, with `out = "matches"`) Boolean matrix indicating which ranges/sequences contain which motifs, stored as "matches" in assays slot of SummarizedExperiment object

2) (`out = "scores"`) Same as (1) plus two additional assays -- a matrix with the score of the high motif score within each range/sequence (score only reported if match present) and a matrix with the number of motif matches.

3) (`out = "positions"`) A GenomicRangesList with the ranges of all matches within the input ranges/sequences. 

## Quickstart

```r
library(motifmatchr)
library(GenomicRanges)

# load some example motifs
data(example_motifs, package = "motifmatchr") 

# Make a set of peaks
peaks <- GRanges(seqnames = c("chr1","chr2","chr2"),
                 ranges = IRanges(start = c(76585873,42772928,100183786),
                                  width = 500))

# Get motif matches for example motifs in peaks
motif_ix <- matchMotifs(example_motifs, peaks, genome = "hg19") 
motifMatches(motif_ix) # Extract matches matrix from SummarizedExperiment result

# Get motif positions within peaks for example motifs in peaks 
motif_ix <- matchMotifs(example_motifs, peaks, genome = "hg19",
                         out = "positions") 
```

## More information

For a more detailed overview, see [vignette](https://greenleaflab.github.io/motifmatchr/articles/motifmatchr.html).


inst/NEWS
# motifmatchr 0.99.7

API changed from snake_case to camelCase. snake_case versions of methods
kept with deprecation notice.

# motifmatchr 0.99.6

Genome no longer defaults to hg19 when subject is GenomicRanges or
RangedSummarizedExperiment, but tries to pull assoicated genome build from
object. Genome can be provided as BSgenome, FaFile, DNAStringSet object, or
short string (e.g. "hg19").

Added BSgenomeViews method.





R/match_motifs.R
#' motif_matches
#'
#' get motif matches from SummarizedExperiment object
#' @param object SummarizedExperiment object with matches assay
#' @return matrix with scores
#' @export
#' @examples
#'
#' data(example_motifs, package = "motifmatchr")
#'
#' # Make a set of peaks
#' peaks <- GenomicRanges::GRanges(seqnames = c("chr1","chr2","chr2"),
#'                 ranges = IRanges::IRanges(start = c(76585873,42772928,
#'                                           100183786),
#'                                           width = 500))
#'
#' # Get motif matches for example motifs
#' motif_ix <- match_motifs(example_motifs, peaks, genome = "hg19")
#'
#' motif_matches(motif_ix)
setGeneric("motif_matches", function(object) standardGeneric("motif_matches"))

#' @describeIn motif_matches method for SummarizedExperiment
#' @export
setMethod("motif_matches", c(object = "SummarizedExperiment"),
          function(object) {
              if ("motif_matches" %ni% assayNames(object))
                  stop("No motif_matches in this object")
              assays(object)$motif_matches
          })


#' motif_scores
#'
#' get motif scores from SummarizedExperiment object
#' @param object SummarizedExperiment object with scores assay
#' @return matrix with scores
#' @export
#' @examples
#'
#' data(example_motifs, package = "motifmatchr")
#'
#' # Make a set of peaks
#' peaks <- GenomicRanges::GRanges(seqnames = c("chr1","chr2","chr2"),
#'                 ranges = IRanges::IRanges(start = c(76585873,42772928,
#'                                           100183786),
#'                                           width = 500))
#'
#' # Get motif matches for example motifs
#' motif_ix <- match_motifs(example_motifs, peaks, genome = "hg19",
#'                          out = "scores")
#'
#' motif_scores(motif_ix)
setGeneric("motif_scores", function(object) standardGeneric("motif_scores"))

#' @describeIn motif_scores method for SummarizedExperiment
#' @export
setMethod("motif_scores", c(object = "SummarizedExperiment"), function(object) {
    if ("motif_scores" %ni% assayNames(object))
        stop("No motif_scores in this object")
    assays(object)$motif_scores
})


#' motif_counts
#'
#' get motif counts from SummarizedExperiment object
#' @param object SummarizedExperiment object with counts assay
#' @return matrix with counts
#' @export
#' @examples
#'
#' data(example_motifs, package = "motifmatchr")
#'
#' # Make a set of peaks
#' peaks <- GenomicRanges::GRanges(seqnames = c("chr1","chr2","chr2"),
#'                 ranges = IRanges::IRanges(start = c(76585873,42772928,
#'                                           100183786),
#'                                           width = 500))
#'
#' # Get motif matches for example motifs
#' motif_ix <- match_motifs(example_motifs, peaks, genome = "hg19",
#'                          out = "scores")
#'
#' motif_counts(motif_ix)
setGeneric("motif_counts", function(object) standardGeneric("motif_counts"))

#' @describeIn motif_counts method for SummarizedExperiment
#' @export
setMethod("motif_counts", c(object = "SummarizedExperiment"), function(object) {
    if ("motif_counts" %ni% assayNames(object))
        stop("No motif_counts in this object")
    assays(object)$motif_counts
})

match_motifs_helper <- function(pwms, seqs, bg, p.cutoff, w, out, ranges) {

    motif_mats <- convert_pwms(pwms, bg)

    if (out == "matches") {
        tmp_out <- get_motif_ix(motif_mats, seqs, bg, p.cutoff, w)
        if (is.null(ranges)) {
            out <- SummarizedExperiment(assays =
                                            list(motif_matches = as(tmp_out,
                                                              "lMatrix")),
                                        colData =
                                            DataFrame(pwm = pwms,
                                                      name = name(pwms),
                                                      row.names = names(pwms)))
        } else {
            out <- SummarizedExperiment(assays =
                                            list(motif_matches = as(tmp_out,
                                                              "lMatrix")),
                                        rowRanges = ranges,
                                        colData =
                                            DataFrame(pwm = pwms,
                                                      name = name(pwms),
                                                      row.names = names(pwms)))
        }
    } else if (out == "scores") {
        tmp_out <- get_motif_ix_plus(motif_mats, seqs, bg, p.cutoff, w)
        tmp_out$motif_matches <- as(tmp_out$motif_matches, "lMatrix")
        if (is.null(ranges)) {
            out <- SummarizedExperiment(assays = tmp_out,
                                        colData =
                                            DataFrame(pwm = pwms,
                                                      name = name(pwms),
                                                      row.names = names(pwms)))
        } else {
            out <- SummarizedExperiment(assays = tmp_out,
                                        rowRanges = ranges,
                                        colData =
                                            DataFrame(pwm = pwms,
                                                      name = name(pwms),
                                                      row.names = names(pwms)))
        }
    } else {
        tmp_out <- get_motif_positions(motif_mats, seqs, bg, p.cutoff, w)
        if (is.null(ranges)) {
            out <- lapply(1:length(motif_mats), function(x) {
                m_ix <- which(tmp_out$motif_ix == x - 1)
                IRangesList(lapply(seq_along(seqs),
                       function(y){
                           ms_ix <- m_ix[which(tmp_out$seq_ix[m_ix] == y -1)]
                           tmp <- IRanges(start = tmp_out$pos[ms_ix] + 1,
                               width = ncol(motif_mats[[x]]))
                            mcols(tmp) <- DataFrame(strand = tmp_out$strand[ms_ix],
                                        score = tmp_out$score[ms_ix])
                            tmp
                       }))
            })
            names(out) <- names(pwms)
        } else {
            out <- lapply(1:length(motif_mats), function(x) {
                m_ix <- which(tmp_out$motif_ix == x - 1)
                GRanges(seqnames(ranges)[tmp_out$seq_ix[m_ix] +  1],
                        IRanges(start =
                                    start(ranges[tmp_out$seq_ix[m_ix] + 1]) +
                                    tmp_out$pos[m_ix],
                                width = ncol(motif_mats[[x]])),
                        strand = tmp_out$strand[m_ix],
                        score = tmp_out$score[m_ix])
            })
            names(out) <- names(pwms)
            out <- GRangesList(out)
        }
    }
    return(out)
}



#' match_motifs
#'
#' Find motif matches
#' @param pwms either \code{\link[TFBSTools]{PFMatrix}},
#' \code{\link[TFBSTools]{PFMatrixList}}, \code{\link[TFBSTools]{PWMatrix}},
#' \code{\link[TFBSTools]{PWMatrixList}}
#' @param subject either \code{\link[GenomicRanges]{GenomicRanges}},
#' \code{\link[Biostrings]{DNAStringSet}}, \code{\link[Biostrings]{DNAString}},
#' or character vector
#' @param genome BSgenome object, \code{\link[Biostrings]{DNAStringSet}}, or
#' \code{\link[Rsamtools]{FaFile}}, or short string signifying genome build
#' recognized by \code{\link[BSgenome]{getBSgenome}}. Only required if
#' subject is \code{\link[GenomicRanges]{GenomicRanges}} or
#' \code{\link[SummarizedExperiment]{RangedSummarizedExperiment}} or if bg is set
#' to "genome"
#' @param bg background nucleotide frequencies. Default is to compute based on
#' subject, i.e. the specific set of sequences being evaluated. See Details.
#' @param out what to return? see return section
#' @param p.cutoff p-value cutoff for returning motifs
#' @param w parameter controlling size of window for filtration; default is 7
#' @param ranges if subject is not GenomicRanges or RangedSummarizedExperiment,
#'  these ranges can be used to specify what ranges the input sequences
#'  correspond to. These ranges will be incorporated into the
#'  SummarizedExperiment output if out is "matches" or "scores" or will be used
#'  to give absolute positions of motifs if out is "positions"
#' @param ... additional arguments depending on inputs
#' @details Background nucleotide frequencies can be set to "subject" to use the
#' subject sequences or ranges for computing the nucleotide frequencies,
#' "genome" for using the genomice frequencies (in which case a genome must be
#' specified), "even" for using 0.25 for each base, or a numeric vector with A,
#' C, G, and T frequencies.
#' @return Either returns a SummarizedExperiment with a sparse matrix with
#'  values set to TRUE for a match (if out == 'matches'), a
#'  SummarizedExperiment with a matches matrix as well as matrices with the
#'  maximum motif score and total motif counts (if out == 'scores'), or a
#'  \code{\link[GenomicRanges]{GenomicRangesList}} or a list of
#'  \code{\link[IRanges]{IRangesList}} with all the positions of matches
#'  (if out == 'positions')
#' @export
#' @examples
#'
#' data(example_motifs, package = "motifmatchr")
#'
#' # Make a set of peaks
#' peaks <- GenomicRanges::GRanges(seqnames = c("chr1","chr2","chr2"),
#'                 ranges = IRanges::IRanges(start = c(76585873,42772928,
#'                                           100183786),
#'                                           width = 500))
#'
#' # Get motif matches for example motifs
#' motif_ix <- match_motifs(example_motifs, peaks, genome = "hg19")
#'
setGeneric("match_motifs",
           function(pwms, subject, ...) standardGeneric("match_motifs"))

#' @describeIn match_motifs PWMatrixList/DNAStringSet
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrixList",
                                    subject = "DNAStringSet"),
          function(pwms,
                   subject,
                   genome = NULL,
                   bg = c("subject","genome","even"),
                   out = c("matches", "scores", "positions"),
                   p.cutoff = 5e-05, w = 7, ranges = NULL) {
              out <- match.arg(out)

              if (is.numeric(bg)){
                  bg <- check_bg(bg)
              } else{
                  bg_method <- match.arg(bg)
                  bg <- get_bg(bg_method, subject, genome)
              }

              seqs <- as.character(subject)

              match_motifs_helper(pwms, seqs, bg, p.cutoff, w, out, ranges)
          })

#' @describeIn match_motifs PWMatrixList/character
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrixList",
                                    subject = "character"),
          function(pwms,
                   subject,
                   genome = NULL,
                   bg = c("subject","genome","even"),
                   out = c("matches", "scores",  "positions"),
                   p.cutoff = 5e-05, w = 7, ranges = NULL) {

              out <- match.arg(out)

              if (is.numeric(bg)){
                  bg <- check_bg(bg)
              } else{
                  bg_method <- match.arg(bg)
                  bg <- get_bg(bg_method, subject, genome)
              }

              match_motifs_helper(pwms, subject, bg, p.cutoff, w, out, ranges)
          })

#' @describeIn match_motifs PWMatrixList/DNAString
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrixList",
                                    subject = "DNAString"),
          function(pwms,
                   subject,
                   genome = NULL,
                   bg = c("subject","genome","even"),
                   out = c("matches", "scores", "positions"),
                   p.cutoff = 5e-05, w = 7, ranges = NULL) {

              out <- match.arg(out)

              if (is.numeric(bg)){
                  bg <- check_bg(bg)
              } else{
                  bg_method <- match.arg(bg)
                  bg <- get_bg(bg_method, subject, genome)
              }

              seqs <- as.character(subject)

              match_motifs_helper(pwms, seqs, bg, p.cutoff, w, out, ranges)
          })

#' @describeIn match_motifs PWMatrixList/GenomicRanges
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrixList",
                                    subject = "GenomicRanges"),
          function(pwms,
                   subject,
                   genome = GenomeInfoDb::genome(subject),
                   bg = c("subject","genome","even"),
                   out = c("matches", "scores", "positions"),
                   p.cutoff = 5e-05, w = 7) {
              out <- match.arg(out)
              GenomicRanges::strand(subject) <- "+"
              genome <- validate_genome_input(genome)
              seqs <- getSeq(genome, subject)

              if (is.numeric(bg)){
                  bg <- check_bg(bg)
              } else{
                  bg_method <- match.arg(bg)
                  bg <- get_bg(bg_method, seqs, genome)
              }

              seqs <- as.character(seqs)

              match_motifs_helper(pwms, seqs, bg, p.cutoff, w, out, subject)
          })

#' @describeIn match_motifs PWMatrixList/RangedSummarizedExperiment
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrixList",
                                    subject = "RangedSummarizedExperiment"),
          function(pwms, subject,
                   genome = GenomeInfoDb::genome(subject),
                   bg = c("subject","genome","even"),
                   out = c("matches", "scores", "positions"),
                   p.cutoff = 5e-05, w = 7) {
              out <- match.arg(out)
              match_motifs(pwms, rowRanges(subject),
                           genome = genome,
                           bg = bg,
                           out = out,
                           p.cutoff = p.cutoff,
                           w = w)
          })

#' @describeIn match_motifs PWMatrixList/BSGenomeViews
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrixList",
                                    subject = "BSgenomeViews"),
          function(pwms,
                   subject,
                   bg = c("subject","genome","even"),
                   out = c("matches", "scores", "positions"),
                   p.cutoff = 5e-05, w = 7) {

              out <- match.arg(out)
              seqs <- as.character(subject)
              ranges <- BSgenome::granges(subject)

              if (is.numeric(bg)){
                  bg <- check_bg(bg)
              } else{
                  bg_method <- match.arg(bg)
                  bg <- get_bg(bg_method, subject, BSgenome::subject(subject))
              }

              seqs <- as.character(subject)

              match_motifs_helper(pwms, seqs, bg, p.cutoff, w, out, ranges)
          })

### PFMatrixList ---------------------------------------------------------------


#' @describeIn match_motifs PFMatrixList/ANY
#' @export
setMethod("match_motifs", signature(pwms = "PFMatrixList", subject = "ANY"),
          function(pwms,
                   subject,
                   ...) {


              pwms_list <- do.call(PWMatrixList, lapply(pwms, toPWM))
              match_motifs(pwms_list,
                           subject,
                           ...)
          })

# Single PWM input -------------------------------------------------------------

#' @describeIn match_motifs PWMatrix/ANY
#' @export
setMethod("match_motifs", signature(pwms = "PWMatrix", subject = "ANY"),
          function(pwms,
                   subject,
                   ...) {

              pwms_list <- PWMatrixList(pwms)
              match_motifs(pwms_list,
                           subject,
                           ...)
          })


# Single PFM -------------------------------------------------------------------

#' @describeIn match_motifs PFMatrix/ANY
#' @export
setMethod("match_motifs",
          signature(pwms = "PFMatrix", subject = "ANY"),
          function(pwms,
                   subject,
                   ...) {

              pwms_list <- PWMatrixList(toPWM(pwms))
              match_motifs(pwms_list,
                           subject,
                           ...)
          })
